qutip/floquet.py

```diff
@@ -810,191 +812,146 @@ def floquet_markov_mesolve(R, ekets, rho0, tlist, e_ops, f_modes_table=None,
     if opt.tidy:
         R.tidyup()
+    rho0 = rho0.proj() if rho0.isket else rho0
 
-    #
-    # check initial state
-    #
-    if isket(rho0):
-        # Got a wave function as initial state: convert to density matrix.
-        rho0 = ket2dm(rho0)
-
-    #
-    # prepare output array
-    #
-    n_tsteps = len(tlist)
+    # Prepare output object.
     dt = tlist[1] - tlist[0]
-
     output = Result()
     output.solver = "fmmesolve"
     output.times = tlist
-
     if isinstance(e_ops, FunctionType):
-        n_expt_op = 0
         expt_callback = True
-
-    elif isinstance(e_ops, list):
-
-        n_expt_op = len(e_ops)
+        store_states = options.store_states or False
+    else:
         expt_callback = False
-
+        try:
+            e_ops = list(e_ops)
+        except TypeError:
+            raise TypeError("`e_ops` must be iterable or a function") from None
+        n_expt_op = len(e_ops)
         if n_expt_op == 0:
-            output.states = []
+            store_states = True
         else:
-            if not f_modes_table:
-                raise TypeError("The Floquet mode table has to be provided " +
-                                "when requesting expectation values.")
-
             output.expect = []
             output.num_expect = n_expt_op
             for op in e_ops:
-                if op.isherm:
-                    output.expect.append(np.zeros(n_tsteps))
-                else:
-                    output.expect.append(np.zeros(n_tsteps, dtype=complex))
-
+                dtype = np.float64 if op.isherm else np.complex128
+                output.expect.append(np.zeros(len(tlist), dtype=dtype))
+        store_states = options.store_states or (n_expt_op == 0)
+    if store_states:
+        output.states = []
+
+    # Choose which frame transformations should be done on the initial and
+    # evolved states.
+    lab_lookup = [f_modes_table_t, f_energies, T]
+    if (
+        any(x is None for x in lab_lookup)
+        and not all(x is None for x in lab_lookup)
+    ):
+        warnings.warn(
+            "if transformation back to the computational basis in the lab"
+            "frame is desired, all of `f_modes_t`, `f_energies` and `T` must"
+            "be supplied."
+        )
+        f_modes_table_t = f_energies = T = None
+
+    # Initial state.
+    if f_modes_0 is not None:
+        rho0 = rho0.transform(f_modes_0)
+
+    # Evolved states.
+    if floquet_basis:
+        def transform(rho, t):
+            return rho
+    elif f_modes_table_t is not None:
+        # Lab frame, computational basis.
+        def transform(rho, t):
+            f_modes_t = floquet_modes_t_lookup(f_modes_table_t, t, T)
+            f_states_t = floquet_states(f_modes_t, f_energies, t)
+            return rho.transform(f_states_t, True)
+    elif f_modes_0 is not None:
+        # Interaction picture, computational basis.
+        def transform(rho, t):
+            return rho.transform(f_modes_0, False)
```

Comment:
I didn't change this back transformation relative to what you wrote, but please also see comment https://github.com/qutip/qutip/pull/1248#discussion_r576399087.